### PR TITLE
Allow provision for multiple calls to external procedure

### LINKED ISSUE #1820: [SciPy][linalg]: Multiple calls to external procedure throws error
```fortran
subroutine idz_snorm(matvec)
external matvec
call matvec(n,v,m,u,p1,p2,p3,p4)
end

subroutine idz_diffsnorm0(matvec)

external matvec

call matvec(n,v,m,u1,p1,p2,p3,p4)

return
end
```

> LFortran

```console
% lfortran a.f90 --implicit-typing --implicit-interface --show-asr
LCOMPILERS_ASSERT failed: /Users/pgoswami/Desktop/lfortran/src/libasr/asr_utils.h
function set_absent_optional_arguments_to_null(), line number 1914 at 
args.size() == (func->n_args - offset)
zsh: abort      lfortran a.f90 --implicit-typing --implicit-interface --show-asr
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1820.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
TODO:
* [ ] Address [commenthttps://github.com/lfortran/lfortran/pull/1819#pullrequestreview-1484777981)

**Comment by Pranavchiku** [comment]
If I update`external_03`, it still works with GFortran
```fortran

subroutine dfctr()
    implicit none
   external dmprec
   double precision dmprec
   return
end subroutine

program external_03
    implicit none
    call dfctr()
end program
```

**Comment by Pranavchiku** [comment]
Same comment for `external_04`

**Comment by certik** [comment]
GFortran by default does `--implicit-interface` and `--implicit-typing`, we don't. So in GFortran things like these work by default. In light of this, is your question that we should not add these features under the `--implicit-interface`? I think we should, because we are constructing the interface implicitly, so it should not be enabled by default.

**Comment by certik** [comment]
I tested this PR locally. This gives an internal error:
```console
$ lfortran external_04.f90 
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 2097
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/Users/ondrej/repos/lfortran/lfortran/src/bin/lfortran.cpp", line 873
    result = fe.get_asr2(input, lm, diagnostics);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/fortran_evaluator.cpp", line 251
    Result<ASR::TranslationUnit_t*> res2 = get_asr3(*ast, diagnostics);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/fortran_evaluator.cpp", line 273
    compiler_options.symtab_only, compiler_options);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_to_asr.cpp", line 98
    auto res = body_visitor(al, ast, diagnostics, unit, compiler_options,
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 2932
    b.is_body_visitor = true;
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 133
    visit_ast(*x.m_items[i]);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4768
    void visit_ast(const ast_t &b) { visit_ast_t(b, self()); }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4726
    case astType::mod: { v.visit_mod((const mod_t &)x); return; }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4776
    void visit_program_unit(const program_unit_t &b) { visit_program_unit_t(b, self()); }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4377
    case program_unitType::Subroutine: { v.visit_Subroutine((const Subroutine_t &)x); return; }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 1640
    transform_stmts(body, x.n_body, x.m_body);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 104
    this->visit_stmt(*m_body[i]);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4813
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4469
    case stmtType::Assign: { v.visit_Assign((const Assign_t &)x); return; }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 1925
    ASR::expr_t *target = ASRUtils::EXPR(tmp);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4864
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/ast.h", line 4531
    case exprType::Compare: { v.visit_Compare((const Compare_t &)x); return; }
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 4721
    tmp = create_ArrayRef(x.base.base.loc, x.m_args, x.n_args, v_expr, v, f2);
  File "/Users/ondrej/repos/lfortran/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 2563
    ASR::expr_t* m_right_expr = ASRUtils::expr_value(a.m_right);
  File "/Users/ondrej/repos/lfortran/lfortran/src/libasr/asr.h", line 41
    LCOMPILERS_ASSERT(is_a<T>(*f));
AssertFailed: is_a<T>(*f)
```

This gives the following error:
```console
$ lfortran external_05.f90 
semantic error: Dummy argument 'matvec' not defined
 --> external_05.f90:1:1 - 5:1
  |
1 |    subroutine idz_snorm(matvec)
  |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
...
  |
5 |    
  | ...^ 
```

And this succeeds:
```console
$ lfortran external_01.f90 
1.00000000000000000e+00
```

But all three cases should say when they encounter the "external matvec" line that "implicit interfaces are not allowed by default, you can enable them with --implicit-interface".

**Comment by Pranavchiku** [comment]
`external_01.f90` succeeds because we have `enorm` declaration present in the ASR. We just call it externally. This test was added long back.

**Comment by Pranavchiku** [comment]
For `external-04,05.f90`, I will do that by tonight my time :)

**Comment by certik** [comment]
> `external_01.f90` succeeds because we have `enorm` declaration present in the ASR. We just call it externally. This test was added long back.

Right, I thought about that. I think that case is fine for now. Later I think we should prohibit it, since you can't call a global function from inside a module or a program. You can only do it by having an explicit interface to it. But for now that's good.

Great, fix the external 04 and 05 and we can merge this.

**Comment by Pranavchiku** [comment]
This PR can have a final review :)

**Comment by Pranavchiku** [comment]
Tests are failing because of `integration_tests/external_01.f90`.

File: integration_tests/external_01.f90 [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
 double precision function enorm(n) result(y)
     double precision, intent(in) :: n
     y = n
     return
 end function
+
+program main
+    double precision, external :: enorm
+    print *, enorm(1.0d0)
+end program
```
Comment (certik) [inline integration_tests/external_01.f90:10]: We need to allow the original test here, as it is needed for SciPy I believe as well.

Comment (certik) [reply integration_tests/external_01.f90]: Fixed in 43b5895fe3101edf012739f28574f3e91f10c9f5.

**Review by certik — APPROVED** [review]
Excellent, thanks for fixing this and adding tests!

I made a few minor improvements and I think this is good to go.

**Pull request merged**